lsp/nls/src/analysis.rs

```diff
@@ -549,230 +545,243 @@ impl PackedAnalysis {
+                initial_env
+            },
+            |analysis| analysis.stdlib_type_context().clone(),
+            |_, _, _| HashMap::new(),
+        )
     }
 
-    /// Inserts a new analysis. If an analysis was already there for the given file id, return the
-    /// overridden analysis, or `None` otherwise.
-    pub fn insert(&mut self, analysis: PackedAnalysis) -> Option<PackedAnalysis> {
-        if analysis.file_id() == self.stdlib_analysis.file_id() {
-            // Panicking there is a bit exaggerated, but it's a bug to re-analyse the stdlib
-            // several times. At least we'll catch it.
-            panic!("tried to insert the stdlib analysis into the registry, but was already there");
-        }
+    /// Inserts a new analysis, overriding any existing analysis with the same file id.
+    ///
+    /// `callback` receives the standard library's environment and type context,
+    /// and constructs the new analysis from them.
+    pub fn insert_with<'a, F>(&'a mut self, callback: F)
+    where
+        F: for<'std> FnOnce(&Environment<'std>, &TypeContext<'std>) -> PackedAnalysis<'std>,
+    {
+        self.with_mut(|slf| {
+            let analysis = callback(slf.initial_term_env, slf.initial_type_ctxt);
+            if analysis.file_id() == slf.stdlib_analysis.file_id() {
+                // Panicking there is a bit exaggerated, but it's a bug to re-analyse the stdlib
+                // several times. At least we'll catch it.
+                panic!(
+                    "tried to insert the stdlib analysis into the registry, but was already there"
+                );
+            }
 
-        self.analyses.insert(analysis.file_id, analysis)
+            slf.analyses.insert(analysis.file_id(), analysis);
+        })
     }
 
+    /// Returns the analysis for a file.
     pub fn get(&self, file_id: FileId) -> Option<&PackedAnalysis> {
-        if file_id == self.stdlib_analysis.file_id() {
-            Some(&self.stdlib_analysis)
+        if file_id == self.borrow_stdlib_analysis().file_id() {
+            Some(self.borrow_stdlib_analysis())
         } else {
-            self.analyses.get(&file_id)
+            self.borrow_analyses().get(&file_id)
         }
     }
 
-    pub fn get_mut(&mut self, file_id: FileId) -> Option<&mut PackedAnalysis> {
-        if file_id == self.stdlib_analysis.file_id() {
-            Some(&mut self.stdlib_analysis)
+    /// Modifies an existing analysis, if it exists.
+    ///
+    /// `callback` receives a reference to the whole analysis registry *without*
+    /// the analysis being modified, along with a mutable reference to the analysis
+    /// to modify.
+    pub fn modify<T, F>(&mut self, file_id: FileId, callback: F) -> Option<T>
+    where
+        F: for<'std> FnOnce(AnalysisRegistryRef<'_, 'std>, &mut PackedAnalysis<'std>) -> T,
+    {
+        self.modify_and_insert(file_id, |reg, analysis| {
+            Ok::<_, ()>((Vec::new(), callback(reg, analysis)))
```

Comment:
If this can never fail, I believe you can be even more precise using [`Infallible`](https://doc.rust-lang.org/std/convert/enum.Infallible.html) instead of `()`.